File: mau_mau/deck.py
import random
from json import JSONEncoder

from mau_mau.card import Card


class Deck:
    player_cards = []
    opponent_cards = []
    deck = []

    @staticmethod
    def generate_mau_mau_deck() -> deck:
        already_generated = set()
        deck = Deck()

        deck.player_cards = Deck.generate_cards(4, already_generated)
        deck.opponent_cards = Deck.generate_cards(4, already_generated)
        deck.deck = Deck.generate_cards(24, already_generated)
        return deck

    @staticmethod
    def generate_cards(number_of_cards: int, already_generated: set) -> list:
        new_cards = []
        while len(new_cards) < number_of_cards and len(already_generated) < 32:
            card = Card(random.randint(0, 8) * 112, random.randint(0, 8) * 187)
            if card not in already_generated:
                new_cards.append(card)
                already_generated.add(card)

        return new_cards


class DeckEncoder(JSONEncoder):
    def default(self, o):
        return o.__dict__

File: frontend/src/data/deck.ts
import {CardTypes} from "@/data/CardTypes";
import {Card} from "@/interfaces/card";

export class Deck {
    public playerCards: Card[] = [];
    public opponentCards: Card[] = [];
    public deck: Card[] = [];
    public playedCards: Card[] = [];

    private static alreadyGenerated: Map<string, Card>;
    private static cardRowValues = [7, 8, 9, 10, 11, 12, 13, 14];
    private static cardColumnTypes = [CardTypes.ACORNS, CardTypes.LEAVES, CardTypes.BELLS, CardTypes.HEARTS];
    private static cardWidth = 112;
    private static cardHeight = 187;

    static generateDeck(): Deck {
        this.alreadyGenerated = new Map();
        const deck = new Deck();

        deck.playerCards = Array.from(this.generateCards(4));
        deck.opponentCards = Array.from(this.generateCards(4));
        deck.deck = Array.from(this.generateCards(23));
        deck.playedCards = Array.from(this.generateCards(1));

        return deck;
    }

    private static generateCards(numberOfCards: number): Set<Card> {
        const newCards: Set<Card> = new Set();
        while (newCards.size < numberOfCards && this.alreadyGenerated.size < 32) {
            let xValue = Math.floor(Math.random() * 8);
            let yType = Math.floor(Math.random() * 4);
            const x = xValue * this.cardWidth;
            const y = yType * this.cardHeight;
            if(xValue == 0) {
                xValue = 8;
            }
            if(yType == 0) {
                yType = 4;
            }
            const card: Card = {
                x: x,
                y: y,
                cardType: this.cardColumnTypes[yType-1],
                value: this.cardRowValues[xValue-1],
            };

            if (!Deck.alreadyGenerated.has(JSON.stringify(card))) {
                newCards.add(card);
                Deck.alreadyGenerated.set(JSON.stringify(card), card);
            }
        }
        return newCards;
    }

}